### ISSUE TITLE
ConnectionMultiplexer opens multiple connections to Sentinel and Standalone

### ISSUE THREAD

**Issue by mnikolaevv4** [StackExchange/StackExchange.Redis#2395]
Hello,

Currently we are using StackExchange.Redis and custom connection pooling implementation that opens multiple connections to Redis as we have high volume of operations and usage in our applications which sometimes caused Timeouts. As of today we are using the following code:

multiplexer = ConnectionMultiplexer.Connect(redisConfiguration.ConfigurationOptions);

We recently updated our StackExchange.Redis version to the latest stable (v2.6.96) and I saw that there is a fix regarding the Endpoints collection which was getting modified on every other Connect method call after the first one. Another change we saw in one of the threads here in Github was that the CommandMap is now handled internally and does not need to be set when using Sentinel. I passed a logger to see what is happening and modified the code like this:

multiplexer = ConnectionMultiplexer.Connect(redisConfiguration.ConfigurationOptions, streamWriter);

What I saw from the logs surprised me. First there is a connection to Sentinel and then to Standalone. However if we use SentinelConnect only one connection to Sentinel gets opened:

multiplexer = ConnectionMultiplexer.SentinelConnect(redisConfiguration.ConfigurationOptions, streamWriter);

The reason we are updating to the latest version and to the usage of **ConnectionMultiplexer.SentinelConnect**  is because when we open more connections we are getting the following error:

**Sentinel: The ConnectionMultiplexer is not a Sentinel connection. Detected as: Standalone**

In our case, we only need Sentinel connection as we do not need multiple Standalone connections being opened. Can you advise us on whether we should use Connect or SentinelConnect and why does Connect opens connections to Sentinel and Standalone?

Here are the logs when using ConnectionMultiplexer.Connect: https://pastebin.com/pLbdVgtH
Here are the logs when using ConnectionMultiplexer.SentinelConnect: https://pastebin.com/Y1Y1y5qy

Regards,
Martin

**Comment by mnikolaevv4** [comment]
Hello Nick,

We are using many connections because during high volume of Redis usage, we were getting Timeout exceptions, which we managed to control by using connection pooling. Currently we are using v2.5.43 so we are not so much behind in the version pipeline. I am sending the configuration that we are currently using:


if (options == null)
                {
                    ConfigurationOptions newOptions;

                    if (!string.IsNullOrEmpty(ConnectionString))
                    {
                        newOptions = ConfigurationOptions.Parse(ConnectionString);
                    }
                    else
                    {
                        newOptions = new ConfigurationOptions()
                        {
                            Ssl = Ssl,
                            AllowAdmin = AllowAdmin,
                            Password = Password,
                            ConnectTimeout = ConnectTimeout,
                            SyncTimeout = SyncTimeout,
                            AbortOnConnectFail = AbortOnConnectFail,
                            ConfigurationChannel = ConfigurationChannel,
                            SslProtocols = sslProtocols,
                            ChannelPrefix = KeyPrefix
                        };

                        if (IsSentinelCluster)
                        {
                            newOptions.ServiceName = ServiceName;
                            //newOptions.CommandMap = CommandMap.Sentinel;

                            foreach (var redisHost in Hosts)
                                newOptions.EndPoints.Add(redisHost.Host, 26379);
                        }
                        else
                        {
                            foreach (var redisHost in Hosts)
                                newOptions.EndPoints.Add(redisHost.Host, redisHost.Port);
                        }
                    }

                    //if (ExcludeCommands != null)
                    //{
                    //    newOptions.CommandMap = CommandMap.Create(
                    //        new System.Collections.Generic.HashSet<string>(ExcludeCommands),
                    //        false);
                    //}

                    if (WorkCount > 0)
                    {
                        newOptions.SocketManager = new SocketManager(GetType().Name, WorkCount,
                            SocketManager.SocketManagerOptions.None);
                    }

                    newOptions.CertificateValidation += CertificateValidation;
                    options = newOptions;
                }
                else
                {
                    if (IsSentinelCluster)
                    {
                        // Reset endpoints and add them again due to StackExchage.Redis overrides of ports to 6379
                        options.EndPoints.Clear();

                        foreach (var redisHost in Hosts)
                            options.EndPoints.Add(redisHost.Host, 26379);
                    }
                }

                return options;
            }

**Comment by mnikolaevv4** [comment]
We are setting AllowAdmin to true, Ssl to false, CommandMap is commented, we are using ServiceName and Password, adding the endpoints and still connections to both Sentinel and Standalone are opened if we use .Connect method instead of .SentinelConnect

**Comment by NickCraver** [comment]
@mnikolaevv4 I'm not 100% sure I follow, but to be explicit: the reconnect issue was fixed in 2.6.66 - definitely recommend updating to latest overall though, as we continue to fix anything people encounter quick as we're able: https://stackexchange.github.io/StackExchange.Redis/ReleaseNotes#2666

There's a lot of stuff here:
- Sentinel connects to the sentinel and then to children so it can tell when the primary changes (from Sentinel pub/sub) so that'll happen for each child...or you can connect to the primary after determining it from Sentinel manually, but you'd want to hook up to the pub/sub notifications to handle things. This isn't a scenario we have any built-in support for (`GetSentinelMasterConnection` assumes 1 connection)
- SocketManager is shared, if you're tuning this down it may do fun things across many multiplexers unless you're spinning up your own pool
- Timeouts themselves seem to be the core issue here - do you have some of those exceptions? It may be that I can advise a bit on the root cause here which we're working around in the first place

**Comment by mnikolaevv4** [comment]
@NickCraver Did you check the difference between the logs, I was wondering why we are connecting to 26379 (Sentinel) and then to 6379 (Standalone) when using .Connect method. My main question here is if we only use the .SentinelConnect method, are we losing the pub/sub. we are currently operating with 3 redis sentinels (1 master, 2 slaves).

**Comment by mnikolaevv4** [comment]
We are currently not experiencing Timeouts by using our connection pooling implementation, so that is not the problem anymore. The problem is when we open more connections using the .Connect method, then we are connecting to 6379 ports even though we specifically override all the endpoints to use 26379, this is where I'm lost.

**Comment by NickCraver** [comment]
`.Connect()` by defaults goes to Sentinel, asks what the primary is, and returns a multiplexer so that primary (which will switch primaries, if there is a change, from listening to the pub/sub on the first connection). So yes, expect 2 connections. It does this whenever a `ServiceName` is provided.

Yes, you are losing pub/sub using `.SentinelConnect()`, but you can do the same thing manually. Here's how it works: https://github.com/StackExchange/StackExchange.Redis/blob/main/src/StackExchange.Redis/ConnectionMultiplexer.Sentinel.cs#L35

What you'd want to do here is SentinelConnect to the sentinel server, then create as many using `SentinelGetMasterAddressByName`, and listen for the pub/sub events in the link above to re-create multiplexers when a primary changes. It's not ideal, but it's not a scenario we support (or are likely to), so that's your best option I can think of.

**Comment by NickCraver** [comment]
@mnikolaevv4 The reason I ask about timeouts is you're adding a lot of complexity to avoid the timeouts...that's likely worth revisiting when you consider the complexity of the workaround, perhaps I could help you resolving the original issue. In the grand scheme, we're solving workaround problems here.

**Comment by mnikolaevv4** [comment]
Yes, I agree but most commonly when we were using 1 ConnectionMultiplexer connection on production environment we were getting Timeouts even though we are trying to use Transactions when setting/sending data towards Redis but we also perform a lot of get operations so you must imagine our troubles. Our implementation is working for more than a year now, so we have some reservations to revert to using only 1 connection instead of a pooled one, but when we open more connections we are getting the error: **The connection multiplexer is not a Sentinel connection.**, that's why I was asking about the difference between .Connect and .SentinelConnect. I still do not understand why do we have connections to 6379. 

Task summary:
09:25:52.4501:   172.17.197.67:26379: Returned with success as Sentinel primary (Source: Already connected)
09:25:52.4501:   172.17.197.68:26379: Returned with success as Sentinel primary (Source: Already connected)
09:25:52.4501:   172.17.197.69:26379: Returned with success as Sentinel primary (Source: Already connected)

09:25:52.4918: Task summary:
09:25:52.4918:   172.17.197.68:6379: Returned with success as Standalone primary (Source: From command: ECHO)
09:25:52.4918:   172.17.197.67:6379: Returned with success as Standalone replica (Source: From command: ECHO)
09:25:52.4918:   172.17.197.69:6379: Returned with success as Standalone replica (Source: From command: ECHO)

**Comment by mnikolaevv4** [comment]
First we are detecting the 3 endpoints as 3 Primary sentinels, and the on the second attempt we connect to them again using 6379 and we receive 1 Primary and 2 Replicas

**Comment by KendoSai** [comment]
Hi @mnikolaevv4, sorry for jumping into your topic but I'm experiencing a Timeout issue which I'm guessing related to our current 1 ConnectionMultiplexer connection. Is it possible if you could share your connection pooling implementation?

It would be very helpful. Thanks

**Comment by mnikolaevv4** [comment]
Hello @KendoSai, we are currently using this solution for connection pooling -> https://github.com/imperugo/StackExchange.Redis.Extensions but we have replicated the code from it so that we can keep StackExchange.Redis updated and maintain the code on our side. We have commented the WorkCount and ExcludeCommands property completely in this class related to the Configuration options -> https://github.com/imperugo/StackExchange.Redis.Extensions/blob/master/src/core/StackExchange.Redis.Extensions.Core/Configuration/RedisConfiguration.cs

**Comment by KendoSai** [comment]
Hi @mnikolaevv4, thanks a lot. Let me try out that way.

**Comment by KendoSai** [comment]
@mnikolaevv4: it seems that I stand in your shoes today. I got exactly the issue you mentioned, from the 2nd connection, it throws error 

Sentinel: The ConnectionMultiplexer is not a Sentinel connection. Detected as: Standalone

![image](https://github.com/StackExchange/StackExchange.Redis/assets/4116423/d01566a6-22c2-4ac4-9954-b1b298bcbbb7)

May I ask if you have any fixes until now? I'm thinking using .SentinelConnect or changing the pool size back to 1 :(

**Comment by mnikolaevv4** [comment]
Hello,

This issue was caused by an old version of Redis where the connection options were overridden after the first Connect but the people responsible for StackExchange.Redis fixed it a few version ago. Are you using the latest version?

**Comment by KendoSai** [comment]
Yes, I'm using the latest version of StackExchange.Redis 2.6.116, my connection options were still unmodified. But I'm still getting the exceptions. Did you get the same issue like me when using the 2.6.96 version? How did you manage to resolve that?
